File: platform/platform-core/fastify/src/lifecycles/03_external_compute.lifecycle.mjs
/**
 * External/Cloud Compute Functions Registration Lifecycle
 *
 * This lifecycle runs BEFORE 04-context-resolver to ensure external compute functions
 * are registered before YamlConfig overwrites are resolved.
 *
 * Auto-loading from directories:
 *   Set EXTERNAL_COMPUTE_DIRS env var (colon-separated paths) to auto-load
 *   *.compute.mjs files from additional directories.
 *
 *   Example: EXTERNAL_COMPUTE_DIRS=/app/cloud-compute:/app/provider-compute
 *
 * Manual registration:
 *   Set server.externalComputeFunctions as an object:
 *   {
 *     "function_name": {
 *       fn: (ctx) => value,  // The compute function
 *       scope: ComputeScope.STARTUP  // or ComputeScope.REQUEST
 *     }
 *   }
 *
 *   Or simply:
 *   {
 *     "function_name": (ctx) => value  // defaults to STARTUP scope
 *   }
 *
 * Cloud providers should import and call registerExternalCompute() to add their functions.
 */

import { readdir } from "fs/promises";
import { existsSync } from "fs";
import { join } from "path";

let ComputeScope;
let HAS_COMPUTE_SCOPE = false;

try {
  const resolver = await import("app_yaml_overwrites");
  ComputeScope = resolver.ComputeScope;
  HAS_COMPUTE_SCOPE = true;
} catch (e) {
  // app-yaml-overwrites not available
}

/**
 * Auto-load compute functions from *.compute.mjs files in a directory.
 *
 * @param {Object} server - Fastify server instance
 * @param {string} baseDir - Directory to scan
 * @param {Object} [logger] - Logger instance
 * @returns {Promise<string[]>} - List of loaded function names
 */
async function autoLoadFromDirectory(server, baseDir, logger = console) {
  if (!existsSync(baseDir)) {
    return [];
  }

  const loaded = [];

  try {
    const files = await readdir(baseDir);
    const computeFiles = files.filter((f) => f.endsWith(".compute.mjs"));

    for (const file of computeFiles) {
      const filepath = join(baseDir, file);
      const funcName = file.replace(".compute.mjs", "");

      try {
        const fileUrl = `file://${filepath}`;
        const module = await import(fileUrl);

        if (typeof module.register !== "function") {
          logger.warn({ filepath }, `[03-external-compute] ${filepath} does not export 'register' function, skipping`);
          continue;
        }

        const name = module.NAME || funcName;
        const scope = module.SCOPE || (HAS_COMPUTE_SCOPE ? ComputeScope.STARTUP : null);

        // Store in externalComputeFunctions for later registration
        server.externalComputeFunctions[name] = {
          fn: module.register,
          scope,
        };
        loaded.push(name);

        logger.info({ name, baseDir }, `[03-external-compute] Loaded from ${baseDir}: ${name}`);
      } catch (e) {
        logger.error({ err: e, filepath }, `[03-external-compute] Error loading ${filepath}: ${e.message}`);
      }
    }
  } catch (e) {
    logger.error({ err: e, baseDir }, `[03-external-compute] Error scanning directory ${baseDir}: ${e.message}`);
  }

  return loaded;
}

/**
 * Register an external compute function to be loaded before YAML resolution.
 *
 * @param {Object} server - Fastify server instance
 * @param {string} name - Function name (used as {{fn:name}} in YAML templates)
 * @param {Function} fn - The compute function - receives ctx object, returns value
 * @param {Object} [scope] - ComputeScope.STARTUP (cached) or ComputeScope.REQUEST (per-call)
 */
export function registerExternalCompute(server, name, fn, scope = null) {
  if (!server.externalComputeFunctions) {
    if (!server.hasDecorator("externalComputeFunctions")) {
      server.decorate("externalComputeFunctions", {});
    }
  }

  // Default to STARTUP scope
  const resolvedScope =
    scope || (HAS_COMPUTE_SCOPE ? ComputeScope.STARTUP : null);

  server.externalComputeFunctions[name] = {
    fn,
    scope: resolvedScope,
  };

  server.log.info({ name }, `[03-external-compute] Registered: ${name}`);
}

/**
 * Register multiple external compute functions at once.
 *
 * @param {Object} server - Fastify server instance
 * @param {Object} functions - Object mapping name -> fn or name -> {fn, scope}
 */
export function registerExternalComputeBatch(server, functions) {
  for (const [name, config] of Object.entries(functions)) {
    if (typeof config === "function") {
      registerExternalCompute(server, name, config);
    } else if (config?.fn) {
      registerExternalCompute(server, name, config.fn, config.scope);
    }
  }
}

/**
 * Lifecycle startup hook.
 * Initialize external compute functions registry and auto-load from directories.
 *
 * Environment variables:
 *   EXTERNAL_COMPUTE_DIRS - Colon-separated paths to scan for *.compute.mjs files
 *
 * Cloud provider modules should have already called registerExternalCompute()
 * before this lifecycle runs, or place files in EXTERNAL_COMPUTE_DIRS.
 */
export async function onStartup(server, config) {
  server.log.info("[lifecycle:external-compute] External compute registration lifecycle starting...");

  try {
    // Initialize the externalComputeFunctions object if not exists
    if (!server.hasDecorator("externalComputeFunctions")) {
      server.log.info("[lifecycle:external-compute] Initializing externalComputeFunctions decorator");
      server.decorate("externalComputeFunctions", {});
    }

    // Auto-load from EXTERNAL_COMPUTE_DIRS environment variable
    const externalDirs = process.env.EXTERNAL_COMPUTE_DIRS;
    server.log.debug({ EXTERNAL_COMPUTE_DIRS: externalDirs || null }, "[lifecycle:external-compute] Checking EXTERNAL_COMPUTE_DIRS");

    if (externalDirs) {
      const dirs = externalDirs.split(":").filter(Boolean);
      server.log.info({ dirs }, "[lifecycle:external-compute] Loading from EXTERNAL_COMPUTE_DIRS");

      for (const dir of dirs) {
        const loaded = await autoLoadFromDirectory(server, dir.trim(), server.log);
        if (loaded.length > 0) {
          server.log.info({ dir, loaded }, `[lifecycle:external-compute] Auto-loaded from ${dir}: ${loaded.join(", ")}`);
        } else {
          server.log.debug({ dir }, "[lifecycle:external-compute] No compute files found in dir");
        }
      }
    }

    // Also check config for additional directories
    const configDirs = config?.external_compute_dirs || config?.externalComputeDirs;
    if (configDirs && Array.isArray(configDirs)) {
      server.log.info({ configDirs }, "[lifecycle:external-compute] Loading from config directories");
      for (const dir of configDirs) {
        const loaded = await autoLoadFromDirectory(server, dir, server.log);
        if (loaded.length > 0) {
          server.log.info({ dir, loaded }, `[lifecycle:external-compute] Auto-loaded from config dir ${dir}: ${loaded.join(", ")}`);
        }
      }
    }

    // Log current state
    const registered = Object.keys(server.externalComputeFunctions || {});
    if (registered.length > 0) {
      server.log.info({ registered }, `[lifecycle:external-compute] Total external functions ready: ${registered.join(", ")}`);
    } else {
      server.log.info("[lifecycle:external-compute] No external compute functions loaded");
    }

    server.log.info("[lifecycle:external-compute] External compute registration lifecycle complete");
  } catch (err) {
    server.log.error({ err, hookName: "03-external-compute" }, "[lifecycle:external-compute] External compute registration failed");
    throw err;
  }
}

File: platform/fastapi_server/config/lifecycle/03_external_compute.lifecycle.py
"""
External/Cloud Compute Functions Registration Lifecycle

This lifecycle runs BEFORE 04_context_resolver to ensure external compute functions
are registered before YamlConfig overwrites are resolved.

Auto-loading from directories:
    Set EXTERNAL_COMPUTE_DIRS env var (colon-separated paths) to auto-load
    *.compute.py files from additional directories.

    Example: EXTERNAL_COMPUTE_DIRS=/app/cloud-compute:/app/provider-compute

Manual registration:
    Set app.state.external_compute_functions as a dict:
    {
        "function_name": {
            "fn": callable,  # The compute function (ctx) -> value
            "scope": ComputeScope.STARTUP  # or ComputeScope.REQUEST
        }
    }

    Or simply:
    {
        "function_name": callable  # defaults to STARTUP scope
    }

Cloud providers should import and call register_external_compute() to add their functions.
"""
import logging
import os
import glob
import importlib.util
from pathlib import Path
from fastapi import FastAPI
from typing import Callable, Dict, Any, Union, List

logger = logging.getLogger("lifecycle:external_compute")

try:
    from app_yaml_overwrites import ComputeScope
    HAS_COMPUTE_SCOPE = True
except ImportError:
    HAS_COMPUTE_SCOPE = False
    ComputeScope = None


def auto_load_from_directory(app: FastAPI, base_dir: str) -> List[str]:
    """
    Auto-load compute functions from *.compute.py files in a directory.

    Args:
        app: FastAPI application instance
        base_dir: Directory to scan

    Returns:
        List of loaded function names
    """
    base_path = Path(base_dir)
    if not base_path.exists():
        logger.debug("Directory does not exist, skipping: %s", base_dir)
        return []

    loaded = []
    pattern = str(base_path / "*.compute.py")

    for filepath in glob.glob(pattern):
        filepath = Path(filepath)
        module_name = filepath.stem
        func_name = module_name.replace(".compute", "")

        try:
            spec = importlib.util.spec_from_file_location(module_name, filepath)
            if spec is None or spec.loader is None:
                logger.warning("Could not load spec for %s", filepath)
                continue

            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)

            if not hasattr(module, "register"):
                logger.warning("%s does not export 'register' function, skipping", filepath)
                continue

            name = getattr(module, "NAME", func_name)
            scope = getattr(module, "SCOPE", ComputeScope.STARTUP if HAS_COMPUTE_SCOPE else None)

            # Store in external_compute_functions for later registration
            app.state.external_compute_functions[name] = {
                'fn': module.register,
                'scope': scope
            }
            loaded.append(name)
            logger.info("Loaded from %s: %s", base_dir, name)

        except Exception as e:
            logger.error("Error loading %s: %s", filepath, e, exc_info=True)

    return loaded


def register_external_compute(
    app: FastAPI,
    name: str,
    fn: Callable[[Dict[str, Any]], Any],
    scope: Any = None
) -> None:
    """
    Register an external compute function to be loaded before YAML resolution.

    Args:
        app: FastAPI application instance
        name: Function name (used as {{fn:name}} in YAML templates)
        fn: The compute function - receives ctx dict, returns value
        scope: ComputeScope.STARTUP (cached) or ComputeScope.REQUEST (per-call)
               Defaults to STARTUP if not specified
    """
    if not hasattr(app.state, 'external_compute_functions'):
        app.state.external_compute_functions = {}

    # Default to STARTUP scope
    if scope is None and HAS_COMPUTE_SCOPE:
        scope = ComputeScope.STARTUP

    app.state.external_compute_functions[name] = {
        'fn': fn,
        'scope': scope
    }
    logger.info("Registered external compute function: %s", name)


def register_external_compute_batch(
    app: FastAPI,
    functions: Dict[str, Union[Callable, Dict[str, Any]]]
) -> None:
    """
    Register multiple external compute functions at once.

    Args:
        app: FastAPI application instance
        functions: Dict mapping name -> fn or name -> {"fn": fn, "scope": scope}
    """
    for name, config in functions.items():
        if callable(config):
            register_external_compute(app, name, config)
        else:
            fn = config.get('fn')
            scope = config.get('scope')
            if fn:
                register_external_compute(app, name, fn, scope)


async def onStartup(app: FastAPI, config: dict):
    """
    Initialize external compute functions registry and auto-load from directories.

    Environment variables:
        EXTERNAL_COMPUTE_DIRS - Colon-separated paths to scan for *.compute.py files

    Cloud provider modules should have already called register_external_compute()
    before this lifecycle runs, or place files in EXTERNAL_COMPUTE_DIRS.
    """
    logger.info("Starting external_compute lifecycle hook...")
    try:
        logger.info("External compute registration lifecycle starting...")

        # Initialize the external_compute_functions dict if not exists
        if not hasattr(app.state, 'external_compute_functions'):
            app.state.external_compute_functions = {}

        # Auto-load from EXTERNAL_COMPUTE_DIRS environment variable
        external_dirs = os.environ.get('EXTERNAL_COMPUTE_DIRS', '')
        logger.debug(
            "Checking env var EXTERNAL_COMPUTE_DIRS: %s",
            "set" if external_dirs else "not set",
        )
        if external_dirs:
            dirs = [d.strip() for d in external_dirs.split(':') if d.strip()]
            logger.info("Loading from EXTERNAL_COMPUTE_DIRS: %s", dirs)

            for dir_path in dirs:
                loaded = auto_load_from_directory(app, dir_path)
                if loaded:
                    logger.info("Auto-loaded from %s: %s", dir_path, loaded)

        # Also check config for additional directories
        config_dirs = config.get('external_compute_dirs') or config.get('externalComputeDirs')
        if config_dirs and isinstance(config_dirs, list):
            logger.debug("Loading from config-defined directories: %s", config_dirs)
            for dir_path in config_dirs:
                loaded = auto_load_from_directory(app, dir_path)
                if loaded:
                    logger.info("Auto-loaded from config dir %s: %s", dir_path, loaded)

        # Log current state
        registered = list(app.state.external_compute_functions.keys())
        if registered:
            logger.info("Total external functions ready: %s", registered)
        else:
            logger.info("No external compute functions loaded")

        logger.info("external_compute lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("external_compute lifecycle hook failed: %s", exc, exc_info=True)
        raise
